WreckEngine/Text.cpp

```diff
@@ -1,198 +1,549 @@
     buffer.bind();
-	for (const auto c : instance.text) {
-
-		const auto glyph = instance.font->glyphs.at(c);
-
-		const auto tx = instance.x + glyph.bearing.x * instance.scale
-			     , ty = instance.y - (glyph.size.y - glyph.bearing.y) * instance.scale;
-		const auto dims = glyph.size * instance.scale;
-		float verts[] = {
-			tx,          ty + dims.y, 0, 0,
-			tx,          ty,          0, 1,
-			tx + dims.x, ty,          1, 1,
-			tx,          ty + dims.y, 0, 0,
-			tx + dims.x, ty,          1, 1,
-			tx + dims.x, ty + dims.y, 1, 0
-		};
-
-		glyph.tex.bind();
-		buffer.subdata(verts, sizeof(verts));
-        GL_CHECK(glDrawArrays(GL_TRIANGLES, 0, 6)); // this works, but ignores the rendering pipeline; BAD
-
-		instance.x += (glyph.advance >> 6) * instance.scale;// the advance is measured in 1/64 pixels, i.e. 1/2^6
-	}
-}
\ No newline at end of file
+    buffer.invalidate();
+    buffer.data(dataSize, dataPtr);
+    buffer.unbind();
+
+    arrayCount = vertices.size();
+    dirtyBuffer = false;
+}
+
+void Text::Renderer::init()
+{
+    shader.program = loadProgram("Shaders/text_v.glsl", "Shaders/text_f.glsl");
+    shader.program.use();
+    shader.sampler  = shader.program.getUniform<GLsampler>("text");
+    shader.cam      = GLresource<mat4> (shader.program, "camera");
+    shader.offset   = shader.program.getUniform<vec2>("offset");
+    shader.scale    = shader.program.getUniform<float>("scale");
+}
+
+void Text::render(Render::MaterialPass* matRenderer)
+{
+    if (Text::active)
+    {
+        renderer.renderer = matRenderer;
+        renderer.draw();
+    }
+}
+
+vec2 Text::getDims(char ch, const FontFace* font, float scale)
+{
+    return getDims(static_cast<uint32_t>(ch), font, scale);
+}
+
+vec2 Text::getDims(uint32_t cp, const FontFace* font, float scale)
+{
+    vec2 dims;
+
+    if (cp == ' ')
+    {
+        dims.x = font->spaceWidth;
+    }
+    else if (cp == '\t')
+    {
+        dims.x = font->spaceWidth * 4;
+    }
+    else if (font->glyphs.count(cp))
+    {
+        const auto& glyph = font->glyphs.at(cp);
+        dims = {glyph.advance, glyph.size.y - glyph.bearing.y * 2};
+    }
+
+    return dims * scale;
+}
+
+vec2 Text::getDims(const std::string& text, const FontFace* font, float scale)
+{
+    int lineCount = 0;
+    return getDims(text, font, scale, lineCount);
+}
+
+vec2 Text::getDims(const std::string& text, const FontFace* font, float scale, int& lineCount)
+{
+    float x = 0, maxX = 0;
+    float firstLineHeight = 0;
+    lineCount = 1;
+
+    for (char cp : text)
+    {
+        if (cp == '\n')
+        {
+            x = 0.0f;
+            ++lineCount;
+        }
+        else if (cp == '\r' || cp == '\b')
+        {
+            continue;
+        }
+        else
+        {
+            auto dims = getDims(cp, font, scale);
+            x += dims.x;
+            if (lineCount == 1)
+                firstLineHeight = fmax(firstLineHeight, dims.y);
```

Comment:
just do: `if(dims.y > firstLineHeight) firstLineHeight = dims.y;`
You can even merge the condition into the enclosing if statement